internal/provider/resource_access_offer_test.go

```diff
@@ -0,0 +1,214 @@
+
+package provider
+
+import (
+	"fmt"
+	"regexp"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-testing/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
+	internaltesting "github.com/juju/terraform-provider-juju/internal/testing"
+)
+
+func TestAcc_ResourceAccessOffer(t *testing.T) {
+	SkipJAAS(t)
+	AdminUserName := acctest.RandomWithPrefix("tfadminuser")
+	ConsumeUserName := acctest.RandomWithPrefix("tfconsumeuser")
+	ReadUserName := acctest.RandomWithPrefix("tfreaduser")
+	userPassword := acctest.RandomWithPrefix("tf-test-user")
+	modelName := acctest.RandomWithPrefix("tf-access-model")
+	offerURL := fmt.Sprintf("admin/%s.appone", modelName)
+
+	resourceName := "juju_access_offer.access_appone_endpoint"
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:                 func() { testAccPreCheck(t) },
+		ProtoV6ProviderFactories: frameworkProviderFactories,
+		Steps: []resource.TestStep{
+			{ // Test username overlap validation
+				Config:      testAccResourceAccessOffer(AdminUserName, ConsumeUserName, ReadUserName, "admin", "admin", "", userPassword, modelName),
+				ExpectError: regexp.MustCompile("appears in.*"),
+			},
+			{ // Create the resource with user as admin
+				Config: testAccResourceAccessOffer(AdminUserName, ConsumeUserName, ReadUserName, "admin", "consume", "read", userPassword, modelName),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "offer_url", offerURL),
+					resource.TestCheckTypeSetElemAttr(resourceName, "admin.*", AdminUserName),
+					resource.TestCheckTypeSetElemAttr(resourceName, "consume.*", ConsumeUserName),
+					resource.TestCheckTypeSetElemAttr(resourceName, "read.*", ReadUserName),
+				),
+			},
+			{ // Change Admin to Consume and Consume to Admin
+				Config: testAccResourceAccessOffer(AdminUserName, ConsumeUserName, ReadUserName, "consume", "admin", "read", userPassword, modelName),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "offer_url", offerURL),
+					resource.TestCheckTypeSetElemAttr(resourceName, "admin.*", ConsumeUserName),
+					resource.TestCheckTypeSetElemAttr(resourceName, "consume.*", AdminUserName),
+					resource.TestCheckTypeSetElemAttr(resourceName, "read.*", ReadUserName),
+				),
+			},
+			{ // Remove user from read permission
+				Config: testAccResourceAccessOffer(AdminUserName, ConsumeUserName, ReadUserName, "consume", "admin", "", userPassword, modelName),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "offer_url", offerURL),
+					resource.TestCheckTypeSetElemAttr(resourceName, "admin.*", ConsumeUserName),
+					resource.TestCheckTypeSetElemAttr(resourceName, "consume.*", AdminUserName),
+					resource.TestCheckNoResourceAttr(resourceName, "read.*"),
+				),
+			},
+			{ // Destroy the resource and validate it can be imported correctly
```

Comment:
Before the last test case, you can consider adding a test to import the offer access.

Reply:
Isn't this already tested with ImportStateVerify: true and ImportState:       true?